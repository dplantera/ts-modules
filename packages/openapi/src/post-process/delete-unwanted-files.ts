import { Folder } from "../folder.js";

export function deleteUnwantedFiles(apiPath: string) {
  Folder.of(apiPath).delete(
    "git_push.sh",
    ".gitignore",
    ".openapi-generator",
    ".npmignore",
    ".openapi-generator-ignore"
  );
}
