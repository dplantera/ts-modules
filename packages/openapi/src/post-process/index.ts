export { createZodPostProcessor, createTsPostProcessor, createSpecProcessor } from "./post-process.js";
