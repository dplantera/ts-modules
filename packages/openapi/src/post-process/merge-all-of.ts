/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-unused-vars */
// noinspection JSUnusedLocalSymbols

import { OpenApiBundled } from "../bundle.js";
import { oas30 } from "openapi3-ts";
import _ from "lodash";
import { z, ZodString } from "zod";
import pointer from "jsonpointer";
import { TopologicalSort } from "topological-sort";
import { isRef } from "@redocly/openapi-core";
// noinspection TypeScriptCheckImport
// eslint-disable-next-line
// @ts-ignore
import mergeJsonSchemas from "merge-json-schemas";

type Node = oas30.SchemaObject | undefined | { [key: string]: Node };
type Context = {
  root: OpenApiBundled;
  visited: Map<string, Node>;
  nodes: Map<Node, Node>;
  edges: Map<Node, Array<Node>>;
  path: { all: Array<Array<string>>; current: Array<string> };
};

export function mergeAllOf(bundled: OpenApiBundled) {
  const mergedAllOf = _.cloneDeep(bundled);
  const { collected, ctx } = findSchemaObjectsWithAllOf(mergedAllOf);
  collected.forEach((s) => doMerge(s, ctx));
  return mergedAllOf;
}
function doMerge(schema: any, ctx: Context) {
  const subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject> =
    schema.allOf ?? [];

  const clonedSchemas = _.cloneDeep(subSchemas);
  const resolvedSchemas = clonedSchemas.map((d) =>
    resolveRef(d, ctx, { deleteRef: false })
  );

  // include dangling properties {allOf: [], danglingA: {}, danglingB: [], danglingC: null, ... }
  const danglingProperties = { ..._.omit(schema, "allOf") };
  resolvedSchemas.unshift(danglingProperties);

  const discriminatedParent = _.cloneDeep(
    resolvedSchemas.toReversed().find((p) => !_.isNil(p.discriminator))
  );
  if (_.isDefined(discriminatedParent) && resolvedSchemas.length <= 2) {
    return schema;
  }

  const merged = mergeSubSchemas(resolvedSchemas, ctx);

  if (_.isDefined(discriminatedParent)) {
    // we set back the ref for parents so we have nice inheritance for java
    const parent = _.isDefined(discriminatedParent["$ref"])
      ? _.pick(discriminatedParent, "$ref")
      : discriminatedParent;
    delete merged["$ref"];
    delete merged.allOf;
    schema.allOf = [parent, merged];
    cleanObj(schema, ["allOf"]);
    return;
  }

  // { allOf: [a, b] } ==> {...a, ...b}
  Object.assign(schema, merged);
  delete schema.allOf;
  delete schema["$ref"];
}

function cleanObj<T extends Record<string, any>>(
  obj: T,
  except: Array<keyof T>
) {
  Object.getOwnPropertyNames(obj).forEach((prop) => {
    if (except.includes(prop)) {
      return;
    }
    delete obj[prop];
  });
}

function mergeSubSchemas(
  _resolvedSchemas: Array<oas30.SchemaObject>,
  ctx: Context
) {
  const firstSchema = _resolvedSchemas[0];
  const subschemas = _resolvedSchemas
    .slice(1)
    .map((d) => resolveRef(d, ctx, { deleteRef: true }));

  const merged = subschemas.reduce((acc, curr) => {
    const merged = mergeJsonSchemas([acc, curr]);
    return Object.assign(acc, merged);
  }, firstSchema);
  return merged;
}

function findSchemaObjectsWithAllOf(bundled: OpenApiBundled) {
  const visited = new Map();
  const nodes = new Map();
  const edges = new Map();
  const path = {
    all: <Array<Array<string>>>[],
    current: <Array<string>>[],
  };
  collectAllOfDsp(bundled, undefined, {
    root: bundled,
    visited,
    nodes,
    path,
    edges,
  });
  const sorted = topoSort(edges);
  const collected = [];
  sorted.forEach((s) => {
    if (!nodes.has(s)) {
      return;
    }
    collected.push(nodes.get(s));
    nodes.delete(s);
  });
  collected.push(...nodes.keys());
  return {
    collected,
    ctx: {
      root: bundled,
      visited,
      nodes,
      path,
      edges,
    },
  };
}

function collectAllOfDsp(node: Node, parent: Node | undefined, ctx: Context) {
  if (
    _.isEmpty(node) ||
    typeof node !== "object" ||
    ctx.visited.has(toNodeId(ctx.path.current))
  ) {
    // leaf
    ctx.path.all.push(_.cloneDeep(ctx.path.current));
    return ctx;
  }
  ctx.visited.set(toNodeId(ctx.path.current), node);

  const key = ctx.path.current.at(-1);
  if (!_.isNil(parent) && _.isNil(key)) {
    return ctx;
  }

  for (const [key, value] of Object.entries(node)) {
    const next: Node = value;
    const nextPath = [...ctx.path.current, key];
    const newParent = isSchema(node) ? node : parent;
    collectAllOfDsp(next, newParent, {
      ...ctx,
      path: { ...ctx.path, current: nextPath },
    });
  }
  const refObj = parseRefObj(node);
  if (refObj.success) {
    const resolved = resolveRef(refObj.data, ctx);
    const children = ctx.edges.get(parent) ?? [];
    children.push(resolved);
    ctx.edges.set(parent, children);
  }

  if (isSchema(node) && !ctx.nodes.has(node)) {
    const children = ctx.edges.get(parent) ?? [];
    children.push(node);
    ctx.edges.set(parent ?? ctx.root, children);

    // collect
    if (node.allOf) {
      ctx.nodes.set(node, node);
    }
  }

  return ctx;
}

function toNodeId(path: Array<string>) {
  return path.join("::");
}

function resolveRef(
  data: { $ref: string } | unknown,
  ctx: Context,
  params?: { deleteRef: boolean }
): typeof params extends { deletedRef: true }
  ? oas30.SchemaObject
  : oas30.SchemaObject & { $ref?: string } {
  if (!isRef(data)) {
    return data as oas30.SchemaObject;
  }
  const ref = data.$ref;
  try {
    const propPath = ref.replace("#/", "/");
    if (params?.deleteRef ?? false) {
      delete (data as any).$ref;
    }
    return pointer.get(ctx.root, propPath);
  } catch {
    throw `Error: could not resolve ref ${ref} in: ${ctx.path.current.join(
      "/"
    )}`;
  }
}

const SchemaObject = z.object({
  discriminator: z.record(z.any()).optional(),
  type: z
    .enum(["integer", "number", "string", "boolean", "object", "null", "array"])
    .optional(),
  allOf: z.array(z.any()).optional(),
  oneOf: z.array(z.any()).optional(),
  anyOf: z.array(z.any()).optional(),
  items: z.record(z.any()).optional(),
  properties: z.record(z.any()).optional(),
});

function isSchema(node: Node) {
  const parsed = SchemaObject.safeParse(node);
  return parsed.success && !_.isEmpty(parsed.data);
}

function parseRefObj(node: unknown) {
  return z.object({ $ref: z.string() }).safeParse(node);
}

function topoSort(edges: Map<Node, Array<Node>>) {
  try {
    const nodes = new Map();
    Array.from(edges.keys()).forEach((parent) => {
      if (parent && !nodes.has(parent)) {
        nodes.set(parent, parent);
      }
    });
    Array.from(edges.values()).forEach((children) => {
      children.forEach((c) => {
        if (c && !nodes.has(c)) {
          nodes.set(c, c);
        }
      });
    });

    const sortOp = new TopologicalSort(nodes);

    for (const [parent, children] of edges.entries()) {
      if (!parent) {
        continue;
      }
      const set = new Set(children);
      Array.from(set).forEach((v) => {
        if (!v) {
          return;
        }
        sortOp.addEdge(parent, v);
      });
    }
    return Array.from(sortOp.sort().keys()).reverse();
  } catch (e) {
    throw "Error: could not create dependency graph";
  }
}
