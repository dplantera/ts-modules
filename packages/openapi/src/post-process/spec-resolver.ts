/* eslint-disable no-inner-declarations,@typescript-eslint/no-explicit-any */
import { oas30 } from "openapi3-ts";
import { OpenApiBundled } from "../bundle.js";
import { isRef } from "@redocly/openapi-core";
import pointer from "jsonpointer";
import { z } from "zod";
import { TopologicalSort } from "topological-sort";
import { _ } from "@dsp/node-sdk";

export module SpecResolver {
  export type Node = oas30.SchemaObject | undefined | { [key: string]: Node };
  export type Context = {
    root: OpenApiBundled;
    visited: Map<string, Node>;
    nodes: Map<Node, Node>;
    edges: Map<Node, Array<Node>>;
    path: { all: Array<Array<string>>; current: Array<string> };
  };

  export type CollectCondition = (schema: oas30.SchemaObject) => boolean;
  export function findSchemaObjectsWith(
    bundled: OpenApiBundled,
    ...conditions: Array<CollectCondition>
  ) {
    /* CONTEXT */
    const visited = new Map();
    const nodes = new Map();
    const edges = new Map();
    const path = {
      all: <Array<Array<string>>>[],
      current: <Array<string>>[],
    };

    /* COLLECT */
    collectDFS(bundled, undefined, conditions, {
      root: bundled,
      visited,
      nodes,
      path,
      edges,
    });

    /* SORT by dependency */
    const sorted = topoSort(edges);
    const collected = [];
    sorted.forEach((s) => {
      if (!nodes.has(s)) {
        return;
      }
      collected.push(nodes.get(s));
      nodes.delete(s);
    });
    collected.push(...nodes.keys());

    return {
      collected,
      ctx: {
        root: bundled,
        visited,
        nodes,
        path,
        edges,
      },
    };
  }

  function collectDFS(
    node: Node,
    parent: Node | undefined,
    visitors: Array<CollectCondition>,
    ctx: Context
  ) {
    if (
      _.isEmpty(node) ||
      typeof node !== "object" ||
      ctx.visited.has(toNodeId(ctx.path.current))
    ) {
      // leaf
      ctx.path.all.push(_.cloneDeep(ctx.path.current));
      return ctx;
    }
    ctx.visited.set(toNodeId(ctx.path.current), node);

    const key = ctx.path.current.at(-1);
    if (!_.isNil(parent) && _.isNil(key)) {
      return ctx;
    }

    for (const [key, value] of Object.entries(node)) {
      const next: Node = value;
      const nextPath = [...ctx.path.current, key];
      const newParent = isSchema(node) ? node : parent;
      collectDFS(next, newParent, visitors, {
        ...ctx,
        path: { ...ctx.path, current: nextPath },
      });
    }
    const refObj = parseRefObj(node);
    if (refObj.success) {
      const resolved = resolveRef(refObj.data, ctx);
      const children = ctx.edges.get(parent) ?? [];
      children.push(resolved);
      ctx.edges.set(parent, children);
    }

    if (isSchema(node) && !ctx.nodes.has(node)) {
      const children = ctx.edges.get(parent) ?? [];
      children.push(node);
      ctx.edges.set(parent ?? ctx.root, children);

      // collect
      if (visitors.some((collectCondition) => collectCondition(node))) {
        ctx.nodes.set(node, node);
      }
    }

    return ctx;
  }

  function toNodeId(path: Array<string>) {
    return path.join("::");
  }

  export function resolveRefNode(
    data: { $ref: string } | unknown,
    ctx: Context,
    params?: { deleteRef: boolean }
  ) {
    if (!isRef(data)) {
      return {
        pointer: undefined,
        resolved: resolveRef(data, ctx, params),
      };
    }
    return {
      pointer: data["$ref"],
      resolved: resolveRef(data, ctx, params),
    };
  }

  export function resolveRef(
    data: { $ref: string } | unknown,
    ctx: Context,
    params?: { deleteRef: boolean }
  ): typeof params extends { deletedRef: true }
    ? oas30.SchemaObject
    : oas30.SchemaObject & { $ref?: string } {
    if (!isRef(data)) {
      return data as oas30.SchemaObject;
    }
    const ref = data.$ref;
    try {
      const propPath = ref.replace("#/", "/");
      if (params?.deleteRef ?? false) {
        delete (data as any).$ref;
      }
      return pointer.get(ctx.root, propPath);
    } catch {
      throw `Error: could not resolve ref ${ref} in: ${ctx.path.current.join(
        "/"
      )}`;
    }
  }

  const SchemaObject = z.object({
    discriminator: z.record(z.any()).optional(),
    type: z
      .enum([
        "integer",
        "number",
        "string",
        "boolean",
        "object",
        "null",
        "array",
      ])
      .optional(),
    allOf: z.array(z.any()).optional(),
    oneOf: z.array(z.any()).optional(),
    anyOf: z.array(z.any()).optional(),
    items: z.record(z.any()).optional(),
    properties: z.record(z.any()).optional(),
  });

  function isSchema(node: Node) {
    const parsed = SchemaObject.safeParse(node);
    return parsed.success && !_.isEmpty(parsed.data);
  }

  function parseRefObj(node: unknown) {
    return z.object({ $ref: z.string() }).safeParse(node);
  }

  function topoSort(edges: Map<Node, Array<Node>>) {
    try {
      const nodes = new Map();
      Array.from(edges.keys()).forEach((parent) => {
        if (parent && !nodes.has(parent)) {
          nodes.set(parent, parent);
        }
      });
      Array.from(edges.values()).forEach((children) => {
        children.forEach((c) => {
          if (c && !nodes.has(c)) {
            nodes.set(c, c);
          }
        });
      });

      const sortOp = new TopologicalSort(nodes);

      for (const [parent, children] of edges.entries()) {
        if (!parent) {
          continue;
        }
        const set = new Set(children);
        Array.from(set).forEach((v) => {
          if (!v) {
            return;
          }
          sortOp.addEdge(parent, v);
        });
      }
      return Array.from(sortOp.sort().keys()).reverse();
    } catch (e) {
      throw "Error: could not create dependency graph";
    }
  }
}
