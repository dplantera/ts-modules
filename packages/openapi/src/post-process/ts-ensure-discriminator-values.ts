import _ from "lodash";
import { SourceFile, StructureKind, SyntaxKind } from "ts-morph";

export function tsEnsureDiscriminatorValues(api: SourceFile) {
  const discriminatedUnions = findDiscriminatedUnions(api);
  const interfaces = api.getChildrenOfKind(SyntaxKind.InterfaceDeclaration);
  discriminatedUnions.forEach(({ node, discriminator }) => {
    const { subTypes, mappingValues, propertyName } = discriminator;
    return subTypes?.map((sub, idx) => {
      const unionName = node.getNameNode().getText();
      if (
        _.isNil(subTypes) ||
        _.isNil(mappingValues) ||
        subTypes.length !== mappingValues?.length
      ) {
        throw `Error: Expected discriminated union ${unionName} to have mappings for all subtypes but found: subTypes: ${JSON.stringify(
          subTypes
        )}, mappings: ${JSON.stringify(mappingValues)}`;
      }
      const mappingValue = mappingValues.at(idx);
      const subType = interfaces.find((i) => i.getNameNode().getText() === sub);
      if (_.isNil(mappingValue) || _.isNil(subType) || _.isNil(propertyName)) {
        throw `Error: Could not resolve discriminator mapping for union '${unionName}' and discriminatorProperty '${
          propertyName ?? "UNKNOWN"
        }' : ${JSON.stringify(mappingValue)}, mappings: ${JSON.stringify(
          subType
        )}`;
      }

      const discriminatorProperty = subType.getProperty(propertyName);
      // ensure discriminatorProperty
      if (_.isNil(discriminatorProperty)) {
        subType.addProperty({
          name: propertyName!,
          kind: StructureKind.PropertySignature,
          type: `'${mappingValue}'` ?? "string",
        });
        return subType;
      }
      const propertyValue = discriminatorProperty.getType();
      // ensure allowed discriminator on unsupported type - only literal or union types expected
      if (!propertyValue.isLiteral() && !propertyValue.isUnion()) {
        discriminatorProperty.setType(`'${mappingValue}'`);
        console.log(
          `fixed missing discriminator value on subtype ${subType}.${propertyName} - setting it with ${mappingValue}`
        );
      }
      // make union if allowed value literal does not match it
      else if (
        propertyValue.isLiteral() &&
        propertyValue.getLiteralValue() !== mappingValue
      ) {
        discriminatorProperty.setType(
          `'${propertyValue.getLiteralValue()}' | '${mappingValue}'`
        );
        console.log(
          `fixed missing discriminator value on subtype ${subType}.${propertyName} - extending ${propertyValue.getText()} with ${mappingValue}`
        );
      }
      // add allowed discriminator to union value if not present
      else if (
        !propertyValue
          .getUnionTypes()
          .some((u) => u.getLiteralValue() === mappingValue)
      ) {
        const unions = propertyValue.getUnionTypes().map((u) => u.getText());
        unions.push(`'${mappingValue}'`);
        const newType = unions.join(" | ");
        discriminatorProperty.setType(newType);
        console.log(
          `fixed missing discriminator value on subtype ${subType}.${propertyName} - extending ${JSON.stringify(
            unions
          )} with ${mappingValue}`
        );
      }
      return subType;
    });
  });
}

export function findMergeSignatureWithAnd(api: SourceFile) {
  return api
    .getDescendantsOfKind(SyntaxKind.VariableDeclaration)
    .flatMap((n) => {
      const isMergeSignatureWithAnd = n.getText().includes(".and");
      if (!isMergeSignatureWithAnd) {
        return [];
      }
      return n
        .getDescendantsOfKind(SyntaxKind.PropertyAccessExpression)
        .filter((f) => f.getName() === "and");
    });
}

function findDiscriminatedUnions(api: SourceFile) {
  return api
    .getDescendantsOfKind(SyntaxKind.TypeAliasDeclaration)
    .flatMap((n) => {
      const tags = n.getJsDocs().flatMap((d) => d.getTags());
      const isDiscriminatedUnion = tags.some(
        (t) => t.getTagNameNode().getText() === "discriminatedUnion"
      );
      if (!isDiscriminatedUnion) {
        return [];
      }
      const propertyName = tags
        .find((t) => t.getTagNameNode().getText() === "discriminatorProperty")
        ?.getCommentText();
      const mappingValues = tags
        .find((t) => t.getTagNameNode().getText() === "discriminatorValues")
        ?.getCommentText();
      const subTypes = tags
        .find((t) => t.getTagNameNode().getText() === "subTypes")
        ?.getCommentText();
      return isDiscriminatedUnion
        ? [
            {
              node: n,
              discriminator: {
                propertyName,
                mappingValues: parseUnionFromJsDoc(mappingValues),
                subTypes: parseUnionFromJsDoc(subTypes),
              },
            },
          ]
        : [];
    });
}

function isDefined<T>(v: T | undefined): v is NonNullable<T> {
  return !_.isNil(v);
}

function parseUnionFromJsDoc(mappingValues: string | undefined) {
  return mappingValues?.split("|").filter(isDefined).map(_.trim);
}
