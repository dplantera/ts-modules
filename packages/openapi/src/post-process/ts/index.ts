export * from "./delete-unwanted-files.js";
export * from "./ts-ensure-discriminator-values.js";
