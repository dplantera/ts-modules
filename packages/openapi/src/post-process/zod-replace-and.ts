import { SourceFile, SyntaxKind } from "ts-morph";
import { findMergeSignatureWithAnd } from "./ts-ensure-discriminator-values.js";

export function zodReplaceAnd(zodApi: SourceFile) {
  const mergeProperties = findMergeSignatureWithAnd(zodApi);
  mergeProperties.forEach((p) => {
    const identifier = p
      .getChildrenOfKind(SyntaxKind.Identifier)
      .filter((id) => id.getText() === "and");
    identifier.forEach((id) => {
      id.replaceWithText("merge");
    });
  });
}
