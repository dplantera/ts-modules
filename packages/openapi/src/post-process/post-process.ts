import { Folder } from "../folder.js";
import { Project, SourceFile, SyntaxKind } from "ts-morph";
import path from "node:path";
import _ from "lodash";
import { tsEnsureDiscriminatorValues } from "./ts-ensure-discriminator-values.js";
import { zodReplaceAnd } from "./zod-replace-and.js";
import { deleteUnwantedFiles } from "./delete-unwanted-files.js";
import { OpenApiBundled } from "../bundle.js";
import { mergeAllOf } from "./merge-all-of.js";

export function postProcessSpec(bundled: OpenApiBundled) {
  return mergeAllOf(bundled);
}

/**
 * Processes resulting models
 * @param outDir File containing all ts interfaces
 */
export function postProcessModels(outDir: string) {
  const out = Folder.of(outDir);
  const tsApiPath = out.makeFilePath("api.ts");
  const zodSchemasPath = out.makeFilePath("zod.ts");

  const project = new Project();
  project.addSourceFileAtPath(tsApiPath);
  project.addSourceFileAtPath(zodSchemasPath);

  const tsApi = project.getSourceFile(path.basename(tsApiPath));
  if (_.isNil(tsApi)) {
    throw `Error: Expected source file for provided path: srcFile: ${tsApiPath}, provided: ${outDir} `;
  }
  tsEnsureDiscriminatorValues(tsApi);

  const zodApi = project.getSourceFile(path.basename(zodSchemasPath));
  if (_.isNil(zodApi)) {
    throw `Error: Expected source file for provided path: srcFile: ${tsApiPath}, provided: ${outDir} `;
  }
  zodReplaceAnd(zodApi);
  zodEnsureUnknownEnumVariant(zodApi);

  project.saveSync();
  deleteUnwantedFiles(tsApiPath);
}

function zodEnsureUnknownEnumVariant(zodApi: SourceFile) {
  const enums = findEnums(zodApi);
  enums.forEach((p) => {
    p.replaceWithText(`${p.getText()}.or(UNKNOWN_SCHEMA)`);
  });
}

export function findEnums(api: SourceFile) {
  return api.getDescendantsOfKind(SyntaxKind.CallExpression).flatMap((n) => {
    // eslint-disable-next-line no-useless-escape
    const isEnumExpression = /\.enum\s*\([\[\]\s\w,_\-"']+\)\s*$/u.test(
      n.getText()
    );
    if (!isEnumExpression) {
      return [];
    }
    return [n];
  });
}
