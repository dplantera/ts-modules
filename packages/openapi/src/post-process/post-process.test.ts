import { bundleOpenapi } from "../bundle.js";
import { postProcessModels, postProcessSpec } from "./post-process.js";
import { mergeAllOf } from "./merge-all-of.js";
import { generateTypescriptAxios } from "../generators/index.js";
import { Folder } from "@dsp/node-sdk";

describe("post process", () => {
  describe("spec", () => {
    test("merge all of", async () => {
      const { parsed } = await bundleOpenapi(
        "test/specs/pets-modular/pets-api.yml",
        postProcessSpec
      );
      const result = mergeAllOf(parsed);
      // const loc = Folder.of("out").writeYml("merged-all-of.yml", result as object);
      expect(result).toMatchSnapshot();
    });
  });

  describe("generated code", () => {
    test("ts types", async () => {
      const { outFile: bundled } = await bundleOpenapi(
        "test/specs/pets-modular/pets-api.yml",
        postProcessSpec
      );
      const outDir = generateTypescriptAxios(bundled, "test/out");
      postProcessModels(outDir);
      const files = Folder.of(bundled).readAllFilesAsString();
      files.forEach((f) =>
        expect(f.content).toMatchSnapshot(`post-process-${f.src}`)
      );
    });
  });
});
