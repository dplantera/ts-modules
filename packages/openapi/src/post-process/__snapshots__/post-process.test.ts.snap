// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post process generated code ts types: post-process-B:\\dev\\js\\ts-modules-remote\\packages\\openapi\\tmp\\bundled-pets-api.yml 1`] = `
"openapi: 3.0.3
info:
  title: My service
  description: My service
  version: 1.0.0
servers:
  - url: https://
paths:
  /v1/pets:
    get:
      summary: Get pets
      operationId: getPets
      responses:
        '200':
          description: Ok response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Pet'
components:
  schemas:
    PetBase:
      type: object
      discriminator:
        propertyName: type
      required:
        - type
      properties:
        type:
          type: string
        name:
          type: string
    DogEnum:
      type: string
      enum:
        - DOG_A
        - DOG_B
    Dog:
      allOf:
        - type: object
          discriminator:
            propertyName: type
          required:
            - type
          properties:
            type:
              type: string
            name:
              type: string
        - type: object
          discriminator:
            propertyName: type
          required:
            - type
          properties:
            type:
              type: string
            name:
              type: string
            fluffy:
              type: boolean
            breed:
              $ref: '#/components/schemas/DogEnum'
    Cat:
      allOf:
        - type: object
          discriminator:
            propertyName: type
          required:
            - type
          properties:
            type:
              type: string
            name:
              type: string
        - type: object
          discriminator:
            propertyName: type
          required:
            - type
          properties:
            type:
              type: string
            name:
              type: string
            breed:
              type: string
              enum:
                - CAT_A
                - CAT_B
    Pet:
      oneOf:
        - $ref: '#/components/schemas/Cat'
        - $ref: '#/components/schemas/Dog'
      discriminator:
        propertyName: type
        mapping:
          DOG: '#/components/schemas/Dog'
          CAT: '#/components/schemas/Cat'
"
`;

exports[`post process spec merge all of 1`] = `
{
  "components": {
    "schemas": {
      "Cat": {
        "allOf": [
          {
            "discriminator": {
              "propertyName": "type",
            },
            "properties": {
              "breed": {
                "enum": [
                  "CAT_A",
                  "CAT_B",
                ],
                "type": "string",
              },
              "name": {
                "type": "string",
              },
              "type": {
                "type": "string",
              },
            },
            "required": [
              "type",
            ],
            "type": "object",
          },
          {
            "discriminator": {
              "propertyName": "type",
            },
            "properties": {
              "breed": {
                "enum": [
                  "CAT_A",
                  "CAT_B",
                ],
                "type": "string",
              },
              "name": {
                "type": "string",
              },
              "type": {
                "type": "string",
              },
            },
            "required": [
              "type",
            ],
            "type": "object",
          },
        ],
      },
      "Dog": {
        "allOf": [
          {
            "discriminator": {
              "propertyName": "type",
            },
            "properties": {
              "breed": {
                "$ref": "#/components/schemas/DogEnum",
              },
              "fluffy": {
                "type": "boolean",
              },
              "name": {
                "type": "string",
              },
              "type": {
                "type": "string",
              },
            },
            "required": [
              "type",
            ],
            "type": "object",
          },
          {
            "discriminator": {
              "propertyName": "type",
            },
            "properties": {
              "breed": {
                "$ref": "#/components/schemas/DogEnum",
              },
              "fluffy": {
                "type": "boolean",
              },
              "name": {
                "type": "string",
              },
              "type": {
                "type": "string",
              },
            },
            "required": [
              "type",
            ],
            "type": "object",
          },
        ],
      },
      "DogEnum": {
        "enum": [
          "DOG_A",
          "DOG_B",
        ],
        "type": "string",
      },
      "Pet": {
        "discriminator": {
          "mapping": {
            "CAT": "#/components/schemas/Cat",
            "DOG": "#/components/schemas/Dog",
          },
          "propertyName": "type",
        },
        "oneOf": [
          {
            "$ref": "#/components/schemas/Cat",
          },
          {
            "$ref": "#/components/schemas/Dog",
          },
        ],
      },
      "PetBase": {
        "discriminator": {
          "propertyName": "type",
        },
        "properties": {
          "name": {
            "type": "string",
          },
          "type": {
            "type": "string",
          },
        },
        "required": [
          "type",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "My service",
    "title": "My service",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/v1/pets": {
      "get": {
        "operationId": "getPets",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Ok response",
          },
        },
        "summary": "Get pets",
      },
    },
  },
  "servers": [
    {
      "url": "https://",
    },
  ],
}
`;
